docs/tutorials/contracts/guest-book.md

```diff
@@ -0,0 +1,993 @@
+
+```ts 
+  afterEach(() => {
+    while(messages.length > 0) {
+      messages.pop();
+    }
+  });
+```
+
+Our use of `afterEach` deletes the message we add in each subsequent test.
+
+The second testing block, `attached deposit tests`, makes heavy use of `VMContext` to allow us to check that any attached deposits to a message get successfully added to our account.  
+
+
+### Running Your Tests
+
+In your testing terminal type:
+
+```
+  yarn test
+```
+
+Once finished, you should see passing tests that look like the following:
+
+![Guest Book Contract Unit Test](/docs/assets/guest-book_as_unit-test.jpg)
+
+Well done! We created a smart contract and tested it. 
+
+Write a few more simple tests of your own, and see if we missed any edge cases.
+
+Otherwise, let's deploy our smart contract and interact with it from the terminal.
+
+
+## Contract: Deploy
+
+We deploy our smart contract to NEAR _TestNet_, with either an auto-generated account id (dev deploy) or a permanent one we've already set up. 
+
+
+### Dev Deploy
+
+Every smart contract in NEAR has its [own associated account][NEAR accounts].
+
+```
+  yarn dev
+```
+
+The above deploys your smart contract to NEAR _TestNet_ with a throwaway account. It also starts a dev server, which we will need when we work on our front end. 
+
+If you want to start interacting with your contracts without starting a local server, you can simply run:
+
+```
+  near dev-deploy
+```
+
+It does pretty much the same thing as `yarn dev`, but without the dev server.  So it will generate an account for your contract, `dev-some-autogenerated-account`.
+
+If you have already configured permanent account for your contract, you can deploy it with the following command: 
+
+
+### Prod Deploy 
+
+``` 
+  near deploy \
+  --contractName=permanent-contract-accountid-you-set-up.testnet \
+  --keyPath=./path-to/testnet/permanent-contract-accountid-you-set-up.testnet.json \
+  --wasmFile=./out/main.wasm
+```
+
+
+### Invoking Contract Methods
+
+
+The following command calls our main change method, `addMessage`.
+
+```
+ near call dev-some-autogenerated-account addMessage '{"text": "hello guestbook"}' --accountId someValidNEARAccount.testnet
+```
+
+then run:
+
+```
+ near call dev-some-autogenerated-account getMessages --accountId someValidNEARAccount.testnet
+```
+
+Your terminal output should resemble this:
+
+```
+[
+  {
+    premium: false,
+    sender: 'someValidNEARAccount.testnet',
+    text: 'hello guestbook'
+  }
+]
+```
+
+Fantastic! Our backend is ready to send and receive. Let's build our UI!
+
+### Switching to a permanent account
+
+When you're ready to switch over from a dev account to a permanent one, here's how:
+
+##### Step 1: Create an account for the contract
+
+Visit [NEAR Wallet] and make a new account. You'll be deploying these smart contracts to this new account.
+Now authorize NEAR CLI for this new account, and follow the instructions it gives you:
```

Comment:
Explain what `near login` does

Reply:
Fixed in a030ab61703335d8aa795892ca6dd80473a0a40b